tests/PaymentBundle/Controller/PaymentControllerTest.php

```diff
@@ -0,0 +1,43 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\PaymentBundle\Tests\Controller;
+
+use PHPUnit\Framework\TestCase;
+use Sonata\Component\Basket\BasketFactoryInterface;
+use Sonata\PaymentBundle\Controller\PaymentController;
+
+final class PaymentControllerTest extends TestCase
+{
+    public function testInstanceBasketInterfacePaymentAction(): void
+    {
+        $class = new \ReflectionClass(PaymentController::class);
+        $method = $class->getMethod('getBasketFactory');
+        $method->setAccessible(true);
+
+        $basketFactoryInterface = $this->createMock(BasketFactoryInterface::class);
+
+        $paymentController = $this->createPaymentController($basketFactoryInterface);
+
+        $this->assertSame($basketFactoryInterface, $method->invoke($paymentController));
+    }
+
+    public function createPaymentController($basketFactory = null)
+    {
+        if (null === $basketFactory) {
```

Comment:
`$basketFactory` won't ever be `null` here, since you are explicitly passing a mock object through

Reply:
Yes you right too, but in fact i tried to do another test by calling createPaymentController(null) for testing this part https://github.com/sonata-project/ecommerce/pull/625/files#diff-0c04180e3dc418019a69461490a25e6cR206

but i give up don't know how to test this

Reply:
You will need to create a mock `ContainerInterface` and set it on the controller, in order to test getting the service from the container. Something like this

```php
$container = $this->createMock(ContainerInterface::class);
$container->expects($this->once())
    ->method('get')
    ->with('sonata.basket.factory')
    ->willReturn( $basketFactoryInterface);

$paymentController->setContainer($container);
```